Properly parse RFC 7235 authentication headers (#4249)

Vampire opened the pull request:
fixes #4249

JakeWharton commented on okhttp/src/main/java/okhttp3/Challenge.java:
```diff
@@ -16,67 +16,97 @@
-    this.charset = charset;
+    this.scheme = scheme.toLowerCase();
+    Map<String, String> authParams = new HashMap<>();
+    authParams.put("realm", realm);
```
Can use `Collections.singletonMap`

JakeWharton commented on okhttp/src/main/java/okhttp3/Challenge.java:
```diff
@@ -16,67 +16,97 @@
+    if (authParams == null) {
+      this.authParams = emptyMap();
+    } else {
+      Map<String, String> newAuthParams = new HashMap<>();
```
Please use `LinkedHashMap` to be consistent with the rest of the codebase

JakeWharton commented on okhttp/src/main/java/okhttp3/Challenge.java:
```diff
@@ -16,67 +16,97 @@
+  public Challenge(String scheme, Map<String, String> authParams) {
+    if (scheme == null) throw new NullPointerException("scheme == null");
+    this.scheme = scheme.toLowerCase();
+    if (authParams == null) {
```
Nullable collections are a bit suspect. Do we want to allow this?

It's also worth noting that the lack of `@Nullable` annotations on this parameters means that the package-level `@NonNull` annotation takes precedence and Java callers and Kotlin callers enforcing JSR 305 _cannot_ pass null even if they wanted to.

JakeWharton commented on okhttp/src/main/java/okhttp3/internal/http/HttpHeaders.java:
```diff
@@ -144,53 +175,92 @@ public static Headers varyHeaders(Headers requestHeaders, Headers responseHeader
-        } else {
+      String[] challengeParts = PROXY_AUTHENTICATE_VALUE_SPLIT_PATTERN.split(header);
+      String authScheme = null;
+      Map<String, String> authParams = new HashMap<>();
```
LinkedHashMap again

Vampire commented:
Thx for the comments, I'll look into them, but first I have to solve a bug in my algorithm that the tests revealed. :-(

Vampire commented:
Ok, bug fixed, travis is green, your comments are incorporated. :-)

yschimke commented on okhttp-tests/src/test/java/okhttp3/HeadersTest.java:
```diff
@@ -434,20 +481,285 @@
+    expectedAuthParams.put(null, "abc==");
+    assertEquals(expectedAuthParams, challenges.get(0).authParams());
+
+    // Invalid token68 part and auth params.
```
nit: I know the structure is existing, but each of these seem better suited as individual test methods rather than commented blocks.

Vampire replied:
I totally agree, I just did like always with PRs and tried to adapt to the surroundings.
Want me to put all into separate methods, also the existing ones?

Vampire replied:
Oh, I just noticed the test name which makes my tests in there misplaced anyway. :-D

yschimke commented:
LGTM - Change and tests look very thorough, but I only gave it a cursory review, not against the spec etc.

yschimke commented on okhttp-tests/src/test/java/okhttp3/HeadersTest.java:
```diff
@@ -366,8 +370,14 @@
         .build();
     List<Challenge> challenges = HttpHeaders.parseChallenges(headers, "WWW-Authenticate");
     assertEquals(1, challenges.size());
-    assertEquals("Digest", challenges.get(0).scheme());
```
I'm ok with this, but we should mention this as a potentially breaking change.  I hope most people using challenges are already using them case insensitively.

e.g. this will break

https://github.com/rburgst/okhttp-digest/blob/0205d0ef00273ce800bb012d4fc1cfc72857a3bc/src/test/java/com/burgstaller/okhttp/WrongOrderedAuthenticationHeaderTest.java#L27

swankjesse replied:
I’m curious what the benefit of forcing lowercase is? It costs an allocation.

yschimke replied:
I like this, but it's whether we want a small amount of upgrade pain for what is probably a small set of users.  I'm not as concerned about the allocation, nor turkish i characters :)

Vampire replied:
I'd hope most users stick to the spec and compare case-insentively, so shouldn't hit too many I hope

yschimke commented on okhttp/src/main/java/okhttp3/Challenge.java:
```diff
@@ -16,67 +16,92 @@
+import static java.util.Collections.unmodifiableMap;
 import static okhttp3.internal.Util.ISO_8859_1;
 
 /** An RFC 7617 challenge. */
```
nit: We now refer to this elsewhere as a 7235 challenge.  7617 is Basic auth IIUC

Vampire replied:
You are totally correct.
It was not a pure 7617 challenge before already.
It was any 7235 challenge, but only had their algorithm, realm and charset properties while it now has all their auth params and can be used to implement any auth algorithm.
I changed the comment.

yschimke commented on okhttp/src/main/java/okhttp3/internal/http/HttpHeaders.java:
```diff
@@ -144,53 +181,116 @@ public static Headers varyHeaders(Headers requestHeaders, Headers responseHeader
-        } else if (header.regionMatches(true, matcher.start(1), "charset", 0, 7)) {
-          charset = matcher.group(3);
-        }
+      List<Challenge> headerChallenges = CHALLENGES_CACHE.get(header);
```
The cache makes sense given the repeated nature of the non-preemptive challenges.  But I don't think we should have it in an unbounded form.

I think we should either restrict it in some way e.g. TTL, size.  Or just accept the wasted effort, which is hopefully noise compared to the network cost.

swankjesse replied:
Yeah, I don’t want a ConcurrentHashMap here. The solution is to make decoding challenges fast & simple, not to cache it!

Vampire replied:
So you want me to simply remove the cache?

Vampire replied:
Ok, removed

yschimke commented on okhttp/src/main/java/okhttp3/internal/http/HttpHeaders.java:
```diff
@@ -144,53 +181,116 @@ public static Headers varyHeaders(Headers requestHeaders, Headers responseHeader
-    // challenge   = auth-scheme 1*SP 1#auth-param
-    // realm       = "realm" "=" realm-value
-    // realm-value = quoted-string
     List<Challenge> challenges = new ArrayList<>();
```
TIOLI: is it worth caching and returning an immutable view of the List here?

swankjesse commented on okhttp-tests/src/test/java/okhttp3/HeadersTest.java:
```diff
@@ -366,8 +370,14 @@
-    assertEquals("Digest", challenges.get(0).scheme());
+    assertEquals("digest", challenges.get(0).scheme());
     assertEquals("myrealm", challenges.get(0).realm());
+    Map<String, String> expectedAuthParams = new HashMap<>();
```
nit: we never use HashMap and you shouldn’t either!
https://publicobject.com/2016/02/08/linkedhashmap-is-always-better-than-hashmap/

Vampire replied:
Well, it's only in the tests, but sure, I'll replace it, no problem

swankjesse commented on okhttp-tests/src/test/java/okhttp3/HeadersTest.java:
```diff
@@ -366,8 +370,14 @@
+    expectedAuthParams.put("stale", "FALSE");
+    assertEquals(expectedAuthParams, challenges.get(0).authParams());
 
     // Not strict RFC 2617 header.
```
... for what it’s worth, this really wants to be cut into several tests

Vampire replied:
Ok, I split the tests

swankjesse commented on okhttp-tests/src/test/java/okhttp3/HeadersTest.java:
```diff
@@ -434,43 +482,254 @@
+
+  @Test public void spacesAroundKeyValueSeparator() {
+    Headers headers = new Headers.Builder()
+            .add("WWW-Authenticate", "Basic realm = \"myrealm\"")
```
nit: https://github.com/square/java-code-styles/

Vampire replied:
You mean the indentation I guess? My bad, fixed.

swankjesse commented on okhttp/src/main/java/okhttp3/Challenge.java:
```diff
@@ -16,67 +16,92 @@
+  public Challenge(String scheme, Map<String, String> authParams) {
+    if (scheme == null) throw new NullPointerException("scheme == null");
+    if (authParams == null) throw new NullPointerException("authParams == null");
+    this.scheme = scheme.toLowerCase();
```
I don’t think this is worthwhile, and besides it won’t work in Turkey!
https://en.wikipedia.org/wiki/Dotted_and_dotless_I

Vampire replied:
What do you mean?
The `toLowerCase`?
I don't think there will be auth schemes or auth param names in turkey, do you?
I personally would that accept that as case where one then would have to parse the challenge header themselves.

swankjesse replied:
The issue is that if you have an Android device configured with the default locale as Turkey then "BASIC".toLowerCase() returns "basıc". So you should never use that method without passing a parameter of Locale.US.

swankjesse replied:
https://developer.android.com/reference/java/util/Locale#default_locale

Vampire replied:
Oh, ok, I wasn't aware of that.
Escpecially as it doesn't happen here if I use turkish locale for the `toLowerCase` call.
I added the explicit US locale. now

yschimke replied:
TIL!

swankjesse commented on okhttp/src/main/java/okhttp3/Challenge.java:
```diff
@@ -16,67 +16,92 @@
   }
 
-  private Challenge(String scheme, String realm, Charset charset) {
+  @Deprecated public Challenge(String scheme, String realm) {
```
why deprecate? this is still good, isn’t it?

swankjesse replied:
won’t most callers want an easy two-argument overload?

Vampire replied:
Well, I'd have expected this to be only used by OkHttp internally mostly, but I can undeprecate it of course, was just a suggestion. :-)

Vampire replied:
What about the `withCharset` that I also deprecated?

swankjesse commented on okhttp/src/main/java/okhttp3/Challenge.java:
```diff
@@ -16,67 +16,92 @@
   /** Returns a copy of this charset that expects a credential encoded with {@code charset}. */
-  public Challenge withCharset(Charset charset) {
-    return new Challenge(scheme, realm, charset);
+  @Deprecated public Challenge withCharset(Charset charset) {
```
why deprecate?

Vampire replied:
Well, I thought it is better to use the constructer with charset right away. But I'll undeprecate it, also was just a suggestion.

swankjesse commented on okhttp/src/main/java/okhttp3/Challenge.java:
```diff
@@ -16,67 +16,92 @@
+      } catch (Exception ignore) {
+      }
+    }
+    return ISO_8859_1;
```
is this the official default? I’d prefer ASCII if possible

Vampire replied:
Well, it is what OkHttp already had as default.
Actually it depends on the concrete auth scheme what is the default.
For `Basic` it is the default afair.
Actually I'd say the default should be `null` as it is auth scheme specific, but that will break implementations of `Basic` auth that depend it to be set.

My first thought was to have `Challenge` without any `getRealm` and `getCharset` methods as both are genrally optional and only mandatory by auth scheme and then have a `BasicChallenge` subclass that adds those methods for basic auth. But I tried to do my changes backwards compatibly.

What do you think?

swankjesse commented on okhttp/src/main/java/okhttp3/Challenge.java:
```diff
@@ -16,67 +16,92 @@
-    return scheme
-        + " realm=\"" + realm + "\""
-        + " charset=\"" + charset + "\"";
+    return scheme + " authParams=\"" + authParams + "\"";
```
quoting the auth params seems weird now that they’re in a map

swankjesse commented on okhttp/src/main/java/okhttp3/Response.java:
```diff
@@ -225,10 +225,16 @@ public boolean isRedirect() {
+   * challenges. Otherwise this returns an empty list of challenges.
+   * <p>
+   * If a challenge uses the {@code token68} variant instead of auth params,
+   * there is exactly one auth param in the challenge at key {@code null}.
```
key null doesn’t feel great, though I don’t have better ideas

Vampire replied:
Well, I had in mind either the token as key and a null value or a null key and the token as value or a separate method.
But as it is an exclusive or between the token and the auth params I thought this is the most elegant solution.
If you want me to change it to something else, or maybe a fixed key `token68`, just tell me.

swankjesse commented on okhttp/src/main/java/okhttp3/internal/http/HttpHeaders.java:
```diff
@@ -38,10 +40,45 @@
+          Pattern.compile(OWS_PATTERN_PART + '=' + OWS_PATTERN_PART);
+  private static final Pattern QUOTED_STRING_AUTH_PARAM_AT_END_PATTERN =
+          Pattern.compile(TOKEN_PATTERN_PART + OWS_PATTERN_PART + '=' + OWS_PATTERN_PART
+                  + QUOTED_STRING_PATTERN_PART + '$');
```
The time to init this class just went up ... is there a way to do this with less code?

Vampire replied:
Well, the class init is only done once per JVM run, so I thought this is not too much of an issue, don't you?

swankjesse commented on okhttp/src/main/java/okhttp3/internal/http/HttpHeaders.java:
```diff
@@ -144,53 +181,116 @@ public static Headers varyHeaders(Headers requestHeaders, Headers responseHeader
-   * Parse RFC 7617 challenges, also wrong ordered ones.
-   * This API is only interested in the scheme name and realm.
+   * Parse RFC 7235 challenges.
+   * If a challenge uses the {@code token68} variant instead of auth params,
```
this doc is duplicated from `Response.challenges()`. Since this class isn’t public API, maybe just omit it?

swankjesse commented on okhttp/src/main/java/okhttp3/internal/http/HttpHeaders.java:
```diff
@@ -144,53 +181,116 @@ public static Headers varyHeaders(Headers requestHeaders, Headers responseHeader
-    // realm-value = quoted-string
     List<Challenge> challenges = new ArrayList<>();
     List<String> authenticationHeaders = responseHeaders.values(challengeHeader);
+headerLoop:
```
This all feels like a large solution chasing a small problem.

yschimke commented:
If there are real world examples of bugs with various proxies and services, it makes it more compelling to take on these changes. If the real world already works because noone really implements all this logic and in practice what we have is good enough, the taking on the additional complexity is a tradeoff.

@Vampire do you have any bug reports to show which servers etc are problematic?

Vampire commented:
> My current thought is to land this and then follow up myself to try to cut the implementation down to something I find palatable. Usually that kind of exercise either sheds some code or gives me a better appreciation for the problem being solved.
>
> Would you be cool with that?

Sure, if you come up with better code that passes the tests, be my guest. :-)

> do you have any bug reports to show which servers etc are problematic?

No, unfortunately not.
I just implemented / tested proxy support for Javacord and thus also tested authenticated proxies.
For that I sent valid authentication headers and found that OkHttp doesn't properly parse them as shown in the issue I referenced.
I implemented the proper parsing in Javacord itself currently to make it work with all proxies that might exist and stick to the RFC and thought I'd contribute it upstream, as I also was asked to in the issue. :-)

Vampire commented:
New push with the resolved conversations incorporated is done

yschimke commented:
I think there is an awkward venn diagram of

1) Things obviously valid because of RFC (must?)
2) Things observed from mainstream servers (FANG + Apache + Nginx etc)
3) Things technically valid because of RFC (should?)
4) Things observed with various one off servers e.g. 1 person github implementation of proxies in Ruby

These may conflict, and the priority of OkHttp is probably 1+2 and a simple fast understandable implementation.

It sounds like landing this and iterating on getting it as simple as possible and explicitly deciding to not support things from 3+4 is a valid way forward.  Based on the high quality of the tests you have added, the tests that start failing will be good documentation for sacrifices made.

Vampire commented:
Well, I'm usually all for proper spec support plus supporting broken implementations if necessary.

If you feel like not supporting valid cases for having simpler code, well it's your project and thus your decision. :-)

swankjesse commented:
Also, when in doubt we can see what Firefox & Chrome do. Though OkHttp isn’t a web browser, password challenges are browser-style authentication. That means we probably don’t need to support arbitrary schemes; instead Basıc & Digest are enough.

Server-to-server authentication should almost always be bearer tokens or client certificates and I assume will always be pre-emptive.

I assume server-to-proxy challenges are rarely used in practice.

yschimke commented:
@Vampire you could end up with a choice of implementing the RFC precisely but breaking some major servers.  And complexity to support a theoretical server that doesn't exist is probably the awkward point.

On broken servers, we have often closed okhttp issues suggesting they file issues against the dodgy server.

Vampire commented:
Well, it is only the challenge parsing and that can be done scheme-independently like it is done in my PR.
Which schemes to support e. g. by the `JavaNetAuthenticator` is a different topic.
But with the challenges parsed properly, any `okhttp3.Authenticator` can support any standard or non-standard authentication scheme, as all challgenges are available.
Of course a custom authenticator can also parse the headers themselves, but then there should not be a `challenges()` method in OkHttp, but a `basicChallenges()` and maybe a `digestChallenges()` method that only return the challenges in question. But even then it would probably be good to properly parse the headers like with multiple schemes and so on.

One could e. g. have a corporate proxy that supports `Negotiate` authentication providing SSO through the Windows login and so on.

Vampire commented:
> @Vampire you could end up with a choice of implementing the RFC precisely but breaking some major servers. And complexity to support a theoretical server that doesn't exist is probably the awkward point.

I don't think my changes will break anything, except maybe if someone did not compare the scheme name or auth param keys case-insensitively. Everything else should work like before, plus supporting more valid cases that weren't supported before.

swankjesse commented:
The discussion reminds me of [descriptive vs. prescriptive in language](https://en.m.wikipedia.org/wiki/Linguistic_prescription). I started my career as a prescriptivist, and have begrudgingly grown to be more of a descriptivist!

Specs and software grow apart. Specs are overly broad so they can support unknown future use cases. Software doesn’t follow specs strictly, sometimes to support use cases or constraints that were not anticipated.

yschimke commented:
I'd still suggest landing and then simplifying with good tests in place (Thanks @Vampire!), and possibly decide which to accept breaking.

Vampire commented:
Made some minor adjustments in naming and removed a comment that didn't apply anymore.

swankjesse commented:
Followup: https://github.com/square/okhttp/pull/4359
